service/spi/src/main/java/org/kaazing/gateway/service/collections/MemoryTopic.java

```diff
@@ -0,0 +1,133 @@
+/**
+ * Copyright 2007-2016, Kaazing Corporation. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.kaazing.gateway.service.collections;
+
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.hazelcast.core.ITopic;
+import com.hazelcast.core.Message;
+import com.hazelcast.core.MessageListener;
+import com.hazelcast.monitor.LocalTopicStats;
+import com.hazelcast.monitor.impl.LocalTopicStatsImpl;
+
+public class MemoryTopic<E> extends MemoryDistributedObject implements ITopic<E> {
+
+    private static final ExecutorService AR_EXECUTOR =
+        Executors.newSingleThreadExecutor(runnable -> new Thread(runnable, "memory-topics-add/remove"));
+    private static final ExecutorService P_EXECUTOR =
```

Comment:
I will look into it. The reason for the 2 threads is to avoid deadlocks when add/remove is called from a message listener. These methods are specified as blocking while publish is specified as fire-and-forget. It we make add/remove unblocking it cannot guarantee a happens-before relationship between adding a listener and publishing on a topic.